File: app/lib/core/network/paginated.dart
/// Mirrors the backend `PaginatedResult<T>` shape.
class Paginated<T> {
  const Paginated({
    required this.items,
    required this.total,
    required this.page,
    required this.limit,
    required this.totalPages,
    required this.hasMore,
  });

  factory Paginated.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic>) itemFromJson,
  ) {
    final rawItems = (json['items'] as List<dynamic>? ?? <dynamic>[])
        .whereType<Map<String, dynamic>>()
        .map(itemFromJson)
        .toList();
    return Paginated<T>(
      items: rawItems,
      total: (json['total'] as num?)?.toInt() ?? rawItems.length,
      page: (json['page'] as num?)?.toInt() ?? 1,
      limit: (json['limit'] as num?)?.toInt() ?? rawItems.length,
      totalPages: (json['totalPages'] as num?)?.toInt() ?? 1,
      hasMore: json['hasMore'] as bool? ?? false,
    );
  }

  final List<T> items;
  final int total;
  final int page;
  final int limit;
  final int totalPages;
  final bool hasMore;

  Paginated<T> concat(Paginated<T> next) => Paginated<T>(
        items: [...items, ...next.items],
        total: next.total,
        page: next.page,
        limit: next.limit,
        totalPages: next.totalPages,
        hasMore: next.hasMore,
      );
}

File: ios/SBCContacts/Core/Networking/Paginated.swift
import Foundation

/// Mirrors the backend `PaginatedResult<T>` shape.
struct Paginated<T: Sendable>: Sendable {
    var items: [T]
    var total: Int
    var page: Int
    var limit: Int
    var totalPages: Int
    var hasMore: Bool

    func concat(_ next: Paginated<T>) -> Paginated<T> {
        Paginated(
            items: items + next.items,
            total: next.total,
            page: next.page,
            limit: next.limit,
            totalPages: next.totalPages,
            hasMore: next.hasMore
        )
    }

    func map<U: Sendable>(_ transform: (T) -> U) -> Paginated<U> {
        Paginated<U>(
            items: items.map(transform),
            total: total,
            page: page,
            limit: limit,
            totalPages: totalPages,
            hasMore: hasMore
        )
    }
}

extension Paginated: Decodable where T: Decodable {
    private enum CodingKeys: String, CodingKey {
        case items, total, page, limit, totalPages, hasMore
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let raw = (try? c.decodeIfPresent([Lossy<T>].self, forKey: .items)) ?? []
        items = raw.compactMap(\.value)
        total = c.int(.total) ?? items.count
        page = c.int(.page) ?? 1
        limit = c.int(.limit) ?? items.count
        totalPages = c.int(.totalPages) ?? 1
        hasMore = c.bool(.hasMore) ?? false
    }
}
